src/lib/ecore/efl_promise2.c

```diff
@@ -0,0 +1,277 @@
+   else
+     _efl_future2_dispatch(next_future, next_value, EINA_FALSE);
+}
+
+static Eina_Bool
+_efl_future2_handler_cb(void *data EINA_UNUSED,
+                        int type EINA_UNUSED,
+                        void *event)
+{
+   Efl_Future2 *f = event;
+   _efl_future2_dispatch(f, f->pending_value, f->reject);
+   _pending_futures = eina_list_remove(_pending_futures, f);
+   return ECORE_CALLBACK_DONE;
+}
+
+static void
+_dummy_free(void *user_data EINA_UNUSED, void *func_data EINA_UNUSED)
+{
+}
+
+static Eina_Bool
+_efl_future2_schedule(Efl_Future2 *f, Eina_Value *value, Eina_Bool reject)
+{
+   EINA_SAFETY_ON_TRUE_RETURN_VAL(f->event != NULL, EINA_FALSE);
+
+   f->reject = reject;
+   if (value && value->type == EINA_VALUE_TYPE_PROMISE2)
+     {
+        Efl_Promise2 *p = eina_value_promise2_steal(value);
+        eina_value_free(value);
+        p->future = f;
+     }
+   else
+     {
+        f->event = ecore_event_add(FUTURE_EVENT_ID, f, _dummy_free, NULL);
+        EINA_SAFETY_ON_NULL_RETURN_VAL(f->event, EINA_FALSE);
+        f->pending_value = value;
+        _pending_futures = eina_list_append(_pending_futures, f);
+     }
+   return EINA_TRUE;
+}
+
+Eina_Bool
+efl_promise2_init(void)
+{
+   const char *choice;
+
+   choice = getenv("EINA_MEMPOOL");
+   if ((!choice) || (!choice[0]))
+     choice = "chained_mempool";
+
+   _promise_mp = eina_mempool_add(choice, "Efl_Promise2",
+                                  NULL, sizeof(Efl_Promise2), 16);
+   EINA_SAFETY_ON_NULL_RETURN_VAL(_promise_mp, EINA_FALSE);
+
+   _future_mp = eina_mempool_add(choice, "Efl_Future2",
+                                 NULL, sizeof(Efl_Future2), 16);
+   EINA_SAFETY_ON_NULL_GOTO(_future_mp, err_future);
+
+   FUTURE_EVENT_ID = ecore_event_type_new();
+
+   _future_event_handler = ecore_event_handler_add(FUTURE_EVENT_ID,
+                                                   _efl_future2_handler_cb,
+                                                   NULL);
+   EINA_SAFETY_ON_NULL_GOTO(_future_event_handler, err_handler);
+
+   return EINA_TRUE;
+
+ err_handler:
+   eina_mempool_del(_future_mp);
+   _future_mp = NULL;
+ err_future:
+   eina_mempool_del(_promise_mp);
+   _promise_mp = NULL;
+   return EINA_FALSE;
+}
+
+void
+efl_promise2_shutdown(void)
+{
+   Efl_Future2 *f;
+
+   EINA_LIST_FREE(_pending_futures, f)
+     _efl_future2_dispatch(f, f->pending_value, f->reject);
+
+   ecore_event_handler_del(_future_event_handler);
+   eina_mempool_del(_future_mp);
+   eina_mempool_del(_promise_mp);
+   _promise_mp = NULL;
+   _future_mp = NULL;
+   _future_event_handler = NULL;
+   FUTURE_EVENT_ID = -1;
+}
+
+EAPI Eina_Value *
+efl_promise2_as_value(Efl_Promise2 *p)
+{
+   Eina_Value *v = eina_value_new(EINA_VALUE_TYPE_PROMISE2);
+   EINA_SAFETY_ON_NULL_RETURN_VAL(v, NULL);
+   eina_value_pset(v, p);
+   return v;
+}
+
+EAPI Eina_Value *
+efl_future2_as_value(Efl_Future2 *f)
+{
+   Eina_Value *v = eina_value_new(EINA_VALUE_TYPE_FUTURE2);
```

Comment:
here it would be better to have `EINA_VALUE_TYPE_PROMISE2` with the following code:

```c
static Eina_Value *efl_promise_future_proxy(void *data, const Eina_Value *v) {
   Efl_Promise2 *p = data;
   if (v && v->type != EINA_VALUE_TYPE_ERROR)
      efl_promise2_resolve(p, v);
  else
      efl_promise2_reject(p, v); // you may shortcircuit this using an internal function that doesn't check "v"
}

EAPI Eina_Value *
efl_future2_as_value(Efl_Future 2 *f)
{
   Efl_Promise2 *p = efl_promise_new(NULL);
   efl_future_then(f, efl_promise_future_proxy, efl_promise_future_proxy, efl_promise_future_free, p);
   return eina_promise2_as_value(p);
}
```